File: benchmark/src/main/java/de/hhu/bsinfo/observatory/benchmark/Observatory.java
package de.hhu.bsinfo.observatory.benchmark;

import de.hhu.bsinfo.observatory.benchmark.Benchmark.Mode;
import de.hhu.bsinfo.observatory.benchmark.config.BenchmarkConfig;
import de.hhu.bsinfo.observatory.benchmark.config.OperationConfig;
import de.hhu.bsinfo.observatory.benchmark.config.IterationConfig;
import de.hhu.bsinfo.observatory.benchmark.config.OperationConfig.OperationMode;
import de.hhu.bsinfo.observatory.benchmark.result.Status;
import java.io.BufferedReader;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.lang.reflect.InvocationTargetException;
import java.net.InetSocketAddress;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;

import de.hhu.bsinfo.observatory.generated.BuildConfig;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class Observatory {

    private static final Logger LOGGER = LoggerFactory.getLogger(Observatory.class);

    private final List<Benchmark> benchmarks = new ArrayList<>();

    public Observatory(BenchmarkConfig config, boolean isServer, InetSocketAddress bindAddress, InetSocketAddress remoteAddress) {
        for (OperationConfig operationConfig : config.getOperations()) {
            for(OperationMode mode : operationConfig.getModes()) {
                String operationClassName =
                        "de.hhu.bsinfo.observatory.benchmark." + operationConfig.getName() + "Operation";

                for (IterationConfig iterationConfig : operationConfig.getIterations()) {
                    Benchmark benchmark = instantiateBenchmark(config.getClassName());

                    if (benchmark == null) {
                        return;
                    }

                    Operation operation = null;

                    if (mode == OperationMode.UNIDIRECTIONAL) {
                        operation = instantiateOperation(operationClassName, benchmark,
                                isServer ? Mode.SEND : Mode.RECEIVE, iterationConfig.getCount(),
                                iterationConfig.getSize());
                    } else if (mode == OperationMode.BIDIRECTIONAL) {
                        Operation sendOperation = instantiateOperation(operationClassName, benchmark,
                                Mode.SEND, iterationConfig.getCount(), iterationConfig.getSize());

                        Operation receiveOperation = instantiateOperation(operationClassName, benchmark,
                                Mode.RECEIVE, iterationConfig.getCount(), iterationConfig.getSize());

                        if(sendOperation == null || receiveOperation == null) {
                            return;
                        }

                        if(!(sendOperation instanceof ThroughputOperation) || !(receiveOperation instanceof ThroughputOperation)) {
                            LOGGER.error("Invalid configuration: Only throughput operations may be executed bidirectionally");
                            return;
                        }

                        operation = new BidirectionalThroughputOperation((ThroughputOperation) sendOperation, (ThroughputOperation) receiveOperation);
                    }

                    if (operation == null) {
                        return;
                    }

                    Arrays.stream(config.getParameters())
                            .forEach(parameter -> benchmark.setParameter(parameter.getKey(), parameter.getValue()));

                    benchmark.setServer(isServer);
                    benchmark.setBindAddress(bindAddress);
                    benchmark.setRemoteAddress(remoteAddress);

                    benchmark.addBenchmarkPhase(new InitializationPhase(benchmark));
                    benchmark.addBenchmarkPhase(new ConnectionPhase(benchmark));
                    benchmark.addBenchmarkPhase(new PreparationPhase(benchmark, isServer ? Mode.SEND : Mode.RECEIVE, iterationConfig.getSize()));

                    if(operation.needsFilledReceiveQueue()) {
                        benchmark.addBenchmarkPhase(new FillReceiveQueuePhase(benchmark));
                    }

                    benchmark.addBenchmarkPhase(new WarmUpPhase(benchmark, operation, iterationConfig.getWarmUpIterations()));

                    if(operation.needsFilledReceiveQueue()) {
                        benchmark.addBenchmarkPhase(new FillReceiveQueuePhase(benchmark));
                    }

                    benchmark.addBenchmarkPhase(new OperationPhase(benchmark, operation));

                    benchmark.addBenchmarkPhase(new CleanupPhase(benchmark));

                    benchmarks.add(benchmark);
                }
            }
        }
    }

    public void start() throws InterruptedException {
        for(Benchmark benchmark : benchmarks) {
            LOGGER.info("Executing benchmark '{}'", benchmark.getClass().getSimpleName());

            if(!benchmark.isServer()) {
                Thread.sleep(100);
            }

            if(benchmark.setup() != Status.OK) {
                return;
            }

            benchmark.synchronize();

            benchmark.executePhases();
        }
    }

    private static de.hhu.bsinfo.observatory.benchmark.Benchmark instantiateBenchmark(String className) {
        try {
            Class<?> clazz = Observatory.class.getClassLoader().loadClass(className);
            return (de.hhu.bsinfo.observatory.benchmark.Benchmark) clazz.getConstructor().newInstance();
        } catch (InstantiationException | InvocationTargetException | NoSuchMethodException | IllegalAccessException | ClassNotFoundException e) {
            e.printStackTrace();
            LOGGER.error("Unable to create benchmark of type '{}'", className);

            return null;
        }
    }

    @SuppressWarnings("unchecked")
    private static Operation instantiateOperation(String className, Benchmark benchmark, Mode mode, int operationCount, int operationSize) {
        try {
            Class<? extends Operation> clazz = (Class<? extends Operation>) Class.forName(className);

            return clazz.getDeclaredConstructor(Benchmark.class, Mode.class, int.class, int.class).newInstance(benchmark, mode, operationCount, operationSize);
        } catch (ClassNotFoundException | InstantiationException | IllegalAccessException | InvocationTargetException | NoSuchMethodException e) {
            e.printStackTrace();
            LOGGER.error("Unable to create benchmark phase of type '{}'", className);

            return null;
        }
    }

    public static void printBanner() {
        InputStream inputStream = Observatory.class.getClassLoader().getResourceAsStream("banner.txt");

        if (inputStream == null) {
            return;
        }

        BufferedReader reader = new BufferedReader(new InputStreamReader(inputStream));
        String banner = reader.lines().collect(Collectors.joining(System.lineSeparator()));

        System.out.print("\n");
        System.out.printf(banner, BuildConfig.VERSION, BuildConfig.BUILD_DATE, BuildConfig.GIT_BRANCH, BuildConfig.GIT_COMMIT);
        System.out.print("\n\n");
    }
}

File: cpp/src/benchmark/Observatory.cpp
/*#include <socket-binding/Benchmark.h>
#include <verbs-binding/Benchmark.h>*/

#include <benchmark/phase/ConnectionPhase.h>
#include <benchmark/phase/PreparationPhase.h>
#include <benchmark/phase/CleanupPhase.h>
#include <benchmark/operation/MessagingThroughputOperation.h>
#include <benchmark/operation/RdmaWriteThroughputOperation.h>
#include <benchmark/operation/RdmaReadThroughputOperation.h>
#include <benchmark/operation/MessagingLatencyOperation.h>
#include <benchmark/operation/MessagingPingPongOperation.h>
#include <benchmark/operation/RdmaWriteLatencyOperation.h>
#include <benchmark/operation/RdmaReadLatencyOperation.h>
#include <benchmark/util/BenchmarkFactory.h>
#include <benchmark/util/OperationFactory.h>
#include <benchmark/util/Util.h>
#include <benchmark/phase/InitializationPhase.h>
#include <benchmark/operation/BidirectionalThroughputOperation.h>
#include <benchmark/phase/FillReceiveQueuePhase.h>
#include <benchmark/phase/WarmupPhase.h>
#include <benchmark/phase/OperationPhase.h>
#include <benchmark/BuildConfig.h>
#include "Observatory.h"

#define STRINGIFY(a) #a
#define XSTRINGIFY(a) STRINGIFY(a)

#ifdef OBSERVATORY_BENCHMARK_HEADER_FILE
    #include XSTRINGIFY(OBSERVATORY_BENCHMARK_HEADER_FILE)
#else
    #error OBSERVATORY_BENCHMARK_HEADER_FILE is not set!
#endif

namespace Observatory {

Observatory::Observatory(nlohmann::json &config, std::string &resultPath, bool isServer, int connectionRetries,
                         SocketAddress &bindAddress, SocketAddress &remoteAddress) :
        config(config),
        resultPath(resultPath),
        isServer(isServer),
        connectionRetries(connectionRetries),
        bindAddress(isServer ? bindAddress : SocketAddress(bindAddress.getHostname(), 0)),
        remoteAddress(remoteAddress) {}

void Observatory::registerPrototypes() {
    #ifdef OBSERVATORY_BENCHMARK_CLASS_NAME
        BENCHMARK_REGISTER(OBSERVATORY_BENCHMARK_CLASS_NAME)
    #else
        #error OBSERVATORY_BENCHMARK_CLASS_NAME is not set!
    #endif

    OPERATION_REGISTER(::Observatory::MessagingThroughputOperation)
    OPERATION_REGISTER(::Observatory::RdmaWriteThroughputOperation)
    OPERATION_REGISTER(::Observatory::RdmaReadThroughputOperation)
    OPERATION_REGISTER(::Observatory::MessagingLatencyOperation)
    OPERATION_REGISTER(::Observatory::MessagingPingPongOperation)
    OPERATION_REGISTER(::Observatory::RdmaWriteLatencyOperation)
    OPERATION_REGISTER(::Observatory::RdmaReadLatencyOperation)
}

void Observatory::start() {
    const std::string &benchmarkClassName = BuildConfig::BENCHMARK_CLASS_NAME;

    for(const auto &operationConfig : config["operations"]) {
        for(const auto &mode : operationConfig["modes"]) {
            std::string operationClassName = std::string(operationConfig["name"]).append("Operation");

            for(const auto &iterationConfig : operationConfig["iterations"]) {
                for(uint32_t i = 0; i < operationConfig["repetitions"]; i++) {
                    std::shared_ptr<Benchmark> benchmark;

                    try {
                        benchmark = BenchmarkFactory::newInstance(benchmarkClassName);
                    } catch(std::runtime_error &e) {
                        LOGGER.error("Unable to instantiate benchmark\n\033[0m %s", benchmarkClassName.c_str(), e.what());
                        return;
                    }

                    std::shared_ptr<Operation> operation;

                    try {
                        if(mode == "unidirectional") {
                            operation = OperationFactory::newInstance(operationClassName, benchmark.get(),
                                    isServer ? Benchmark::Mode::SEND : Benchmark::Mode::RECEIVE,iterationConfig["count"], iterationConfig["size"]);
                        } else if(mode == "bidirectional") {
                            std::shared_ptr<Operation> sendOperation = OperationFactory::newInstance(operationClassName, benchmark.get(),
                                    Benchmark::Mode::SEND,iterationConfig["count"],iterationConfig["size"]);
                            std::shared_ptr<Operation> receiveOperation = OperationFactory::newInstance(operationClassName, benchmark.get(),
                                    Benchmark::Mode::RECEIVE,iterationConfig["count"],iterationConfig["size"]);

                            if(!Util::instanceof<ThroughputOperation>(&*sendOperation) || !Util::instanceof<ThroughputOperation>(&*receiveOperation)) {
                                LOGGER.error("Invalid configuration: Only throughput operations may be executed bidirectionally");
                                return;
                            }

                            operation = std::shared_ptr<Operation>(new BidirectionalThroughputOperation(std::static_pointer_cast<ThroughputOperation>(sendOperation), std::static_pointer_cast<ThroughputOperation>(receiveOperation)));
                        }
                    } catch(std::runtime_error &e) {
                        LOGGER.error("Unable to instantiate operation of type '%s'\n\033[0m %s", operationClassName.c_str(), e.what());
                        return;
                    }

                    for(const auto &parameter : config["parameters"]) {
                        benchmark->setParameter(static_cast<std::string>(parameter["key"]).c_str(), static_cast<std::string>(parameter["value"]).c_str());
                    }

                    benchmark->setServer(isServer);
                    benchmark->setConnectionRetries(connectionRetries);

                    benchmark->setDetectorConfig(config["detector"]);

                    benchmark->setBindAddress(bindAddress);
                    benchmark->setRemoteAddress(remoteAddress);

                    benchmark->setResultName(static_cast<std::string>(config["resultName"]).empty() ? static_cast<std::string>(config["className"]).c_str() : static_cast<std::string>(config["resultName"]).c_str());
                    benchmark->setResultPath(static_cast<std::string>(resultPath).c_str());
                    benchmark->setIterationNumber(i);

                    benchmark->addBenchmarkPhase(std::shared_ptr<BenchmarkPhase>(new InitializationPhase(*benchmark)));
                    benchmark->addBenchmarkPhase(std::shared_ptr<BenchmarkPhase>(new ConnectionPhase(*benchmark)));
                    benchmark->addBenchmarkPhase(std::shared_ptr<BenchmarkPhase>(new PreparationPhase(*benchmark, iterationConfig["size"],
                                                                                                      0)));

                    benchmark->addBenchmarkPhase(std::shared_ptr<BenchmarkPhase>(new FillReceiveQueuePhase(*benchmark, *operation)));
                    benchmark->addBenchmarkPhase(std::shared_ptr<BenchmarkPhase>(new WarmupPhase(*benchmark, *operation, static_cast<uint32_t>(iterationConfig["warmUp"]))));

                    benchmark->addBenchmarkPhase(std::shared_ptr<BenchmarkPhase>(new FillReceiveQueuePhase(*benchmark, *operation)));
                    benchmark->addBenchmarkPhase(std::shared_ptr<BenchmarkPhase>(new OperationPhase(*benchmark, *operation)));

                    benchmark->addBenchmarkPhase(std::shared_ptr<BenchmarkPhase>(new CleanupPhase(*benchmark)));

                    executeBenchmark(*benchmark);
                }
            }
        }
    }
}

void Observatory::executeBenchmark(Benchmark &benchmark) {
    Status status = benchmark.setup();
    if(status != Status::OK) {
        exit(status);
    }

    benchmark.executePhases();
}

}